src/lib/signature_lib/checked.ml

```diff
@@ -235,13 +247,33 @@ module Schnorr
     in
     go (Curve.Scalar.length_in_bits - 1) Curve.zero
 
-  let verify ((s, h) : Signature.t) (pk : Public_key.t) (m : Message.t) =
-    let pre_r = shamir_sum (s, Curve.one) (h, pk) in
-    if Curve.equal Curve.zero pre_r then false
-    else
-      let r = compress pre_r in
-      let h' = Message.hash ~nonce:r m in
-      Curve.Scalar.equal h' h
+    open Fold_lib
+
+  (* use with Field.unpack for field elements, Curve.Scalar.to_bits with scalars... *)
+  let to_triples x =
+    Fold.to_list Fold.(group3 ~default:false (of_list x))
+
+  let sign (d : Private_key.t) m =
+    let nonce = to_triples (Curve.Scalar.to_bits d) in
+    let k_prime = Message.hash m ~nonce in
+    assert (not (Curve.Scalar.(equal k_prime zero))) ;
+    let r_point = Curve.scale Curve.one k_prime in
+    let yr = get_y r_point in
+    let k =
+    if is_even yr then k_prime
+    else Curve.Scalar.(order - k_prime)
+    in
+    let nonce = to_triples ((compress r_point) @ compress (Curve.scale Curve.one d)) in
+    let e = Message.hash m ~nonce in
+    let s = Curve.Scalar.(k + e * d) in
+    (get_x r_point, s)
+
+  let verify ((r, s) : Signature.t) (pk : Public_key.t) (m : Message.t) =
+    let nonce = to_triples (Field.unpack r @ (compress pk)) in
+    let e = Message.hash m ~nonce in
+    let r_point = Curve.(subtract (scale one s) (scale pk e)) in
+    (not Curve.(equal zero r_point)) && (is_even (get_y r_point)) &&
```

Comment:
I think it might be cleaners to just have
```ocaml
let (rx, ry) = Curve.to_affine_coordinates r in
```

rather than calling get_x and get_y separately.